pandas/core/frame.py

```diff
@@ -5661,6 +5668,9 @@ def reset_index(
         col_fill : object, default ''
             If the columns have multiple levels, determines how the other
             levels are named. If None then the index name is repeated.
+        allow_duplicates : bool, default False
```

Comment:
so i think we need to default this to lib.no_default (in the signature) that way you can tell if a bool is actually passed (and if its not then you use the flags.allows_duplicate_labels)

Reply:
I am in agreement with @rhshadrach that we need to leave the current behavior alone. That means defaulting to False and ignoring flags.allows_duplicate_labels.

flags.allows_duplicate_labels will always be True, so it is not a useful indication.

As I pointed out above:
> `flags.allows_duplicate_labels = False` seems broken due to inplace handling all over the place (Yes, I will make an Issue).

Reply:
> flags.allows_duplicate_labels will always be True, so it is not a useful indication.

not sure how this would be the case.

-1 on adding this option without full support fo the option itself, otherwise this defeats the purpose

Reply:
Not sure I understand @jreback, for example:

```
df = pd.DataFrame({'a': [1, 2, 3], 'b': [4, 5, 6]}).set_index('a')
df = df.rename(columns={'b': 'a'})
df = df.set_index('a', append=True)

df.reset_index()  # raises `cannot insert a, already exists`
df.reset_index(allow_duplicates=True)  # would be successful
```

This can be useful even without supporting the flag DataFrame.flags.allow_duplicate_labels.

Reply:
this is still not correct. this should use the flags default *unless* a bool is passed, so this *must* default to `lib.no_default`

Reply:
Sorry, but I do not think that it should. The concensus is quite clear and IMO correct, namely that the present behavior should be preserved and the default not changed.
As pointed out elsewhere, the flags value being negative breaks the DataFrame (#44958, https://github.com/pandas-dev/pandas/pull/44755#issuecomment-993694755) so it is no use in its current form
Your suggestion would change the default behavior of reset_index to allow duplicates, something that has been considered (@rhshadrach) to be highly undesirable. Intoducing new duplicate column labels is generally a sign of an upstream error and should raise.

Reply:
>Sorry, but I do not think that it should. The concensus is quite clear and IMO correct, namely that the present behavior should be preserved and the default not changed.

well that is wrong. we have a flag, so if this doesn't use it then its completely sill and we should just entirely scrap it.

if you are adding a allow_duplicates flag to .reset_index then it HAS to respect if the option is set. if there is a bug elsewhere then it should be fixed. 

this PR is breaking the api by doing this.

Reply:
> this PR is breaking the api by doing this.

The API is that `DataFrame.insert()`  already has an `allow_duplicates` parameter that defaults to `False` and ignores `flags.allows_duplicate_labels` https://github.com/pandas-dev/pandas/blob/8e38f280effa2d2ae8417781b2a49c02bd35eb3c/pandas/core/frame.py#L4379
All I am doing is gaining access to that parameter.
BTW, `flags.allows_duplicate_labels` does have an effect, in that if it is False then allow_duplicates=True will raise.

I am about to give up this argument and do it your way but I suspect that I will be vetoed by other reviewers.

Reply:
> if you are adding a allow_duplicates flag to .reset_index then it HAS to respect if the option is set. if there is a bug elsewhere then it should be fixed.

If the allow_duplicates flag is set to False on the frame, this PR will still raise with `.reset_index(allow_duplicates=True)`. The default behavior `.reset_index()` uses `allow_duplicates=False`. When the flag when set to True, it is always is respected.

Working with pandas, difficulty is encountered if one get a duplicate column label. In my experience this can often go unnoticed until later in processing, where it gives rise to mysterious errors. The distance between the introduction of the duplicate label and the error encountered can make it difficult to debug. I would greatly prefer if pandas were to raise when a duplicate _column_ label is introduced regardless of any flag values, but allow the user to override (opt-in only). If the allow_duplicates flag is set to True on the frame, any op should raise regardless of the user specified override.

That is the current behavior of this PR.

Reply:
this is the *entire* point of the flag `allowss_duplicate_labels`

Reply:
i think you all are missing the fact that THE FLAG MATTERS HERE. and you can have it set when this option is in effect.

Reply:
Users whose transformations induce duplicate row labels, but not duplicate column labels, cannot use the flag as it forbids both duplicate row and column labels.

Reply:
> just use the flag. that was my original point

Using the flag was exactly what I tried first. But setting the flag to `False` in testing didn't work, which I have reported (#44958). 

Also, as Richard points out, the flag is for a slightly different purpose.

I have now changed the default to `lib.no_default` and passed the responsibilty down to `DataFrame.insert()`, which is _not_ the subject of this PR. If someone wants to fix the flag and put different logic in `insert()` then that can be done in the future. 

It would be quite wrong to repeat that logic in both `insert()` and `reset_index()`.

For the moment, I have taken pains to _not_ change the default semantics of `reset_index()`.

My use-case, (#44992), is for `Series.reset_index(allow_duplicates=True)`, to allow column label duplicates from the series index.